File: android/posekit/src/test/kotlin/com/oerol/posekit/SimilarityTransformTests.kt
package com.oerol.posekit

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class SimilarityTransformTests {
    private val standing = Fixtures.standing

    @Test
    fun identityRecovered() {
        val t = SimilarityTransform.mapping(standing, standing)!!
        assertEquals(1f, t.scale, 1e-4f)
        assertEquals(0f, t.rotation, 1e-4f)
        assertEquals(0f, t.translation.x, 1e-4f)
        assertEquals(0f, t.translation.y, 1e-4f)
    }

    @Test
    fun knownTransformRecovered() {
        val live = Fixtures.transformed(standing, scale = 1.7f, rotation = 0.4f,
                                        translation = Vec2(0.25f, -0.1f))
        val t = SimilarityTransform.mapping(standing, live)!!
        assertEquals(1.7f, t.scale, 1e-3f)
        assertEquals(0.4f, t.rotation, 1e-3f)
        for ((j, p) in standing.points) {
            val mapped = t.apply(p)
            assertEquals("$j.x", live.points.getValue(j).x, mapped.x, 1e-3f)
            assertEquals("$j.y", live.points.getValue(j).y, mapped.y, 1e-3f)
        }
    }

    @Test
    fun partialOverlapStillMaps() {
        val upper = listOf(
            Joint.nose, Joint.leftEye, Joint.rightEye, Joint.neck,
            Joint.leftShoulder, Joint.rightShoulder, Joint.leftElbow,
            Joint.rightElbow, Joint.leftWrist, Joint.rightWrist, Joint.root,
        )
        val live = Fixtures.transformed(standing, scale = 2.0f, translation = Vec2(1f, 1f))
        val partial = upper.associateWith { live.points.getValue(it) }
        val t = SimilarityTransform.mapping(standing, PoseVector(partial))!!
        val mappedAnkle = t.apply(standing.points.getValue(Joint.leftAnkle))
        assertEquals(live.points.getValue(Joint.leftAnkle).x, mappedAnkle.x, 1e-3f)
        assertEquals(live.points.getValue(Joint.leftAnkle).y, mappedAnkle.y, 1e-3f)
    }

    @Test
    fun tooFewJointsReturnsNull() {
        val tiny = PoseVector(mapOf(Joint.nose to Vec2(0.5f, 0.1f), Joint.neck to Vec2(0.5f, 0.2f)))
        assertNull(SimilarityTransform.mapping(standing, tiny))
    }

    @Test
    fun degenerateReferenceReturnsNull() {
        val p = Joint.entries.associateWith { Vec2(0.5f, 0.5f) }
        assertNull(SimilarityTransform.mapping(PoseVector(p), standing))
    }
}

File: PoseKit/Tests/PoseKitTests/SimilarityTransformTests.swift
import XCTest
@testable import PoseKit

final class SimilarityTransformTests: XCTestCase {
    let standing = Fixtures.standing

    func testIdentityRecovered() {
        let t = SimilarityTransform.mapping(reference: standing, live: standing)!
        XCTAssertEqual(t.scale, 1, accuracy: 1e-4)
        XCTAssertEqual(t.rotation, 0, accuracy: 1e-4)
        XCTAssertEqual(t.translation.x, 0, accuracy: 1e-4)
        XCTAssertEqual(t.translation.y, 0, accuracy: 1e-4)
    }

    func testKnownTransformRecovered() {
        let live = Fixtures.transformed(standing, scale: 1.7, rotation: 0.4,
                                        translation: [0.25, -0.1])
        let t = SimilarityTransform.mapping(reference: standing, live: live)!
        XCTAssertEqual(t.scale, 1.7, accuracy: 1e-3)
        XCTAssertEqual(t.rotation, 0.4, accuracy: 1e-3)
        // The recovered transform must land every reference joint on its live twin.
        for (j, p) in standing.points {
            let mapped = t.apply(p)
            XCTAssertEqual(mapped.x, live.points[j]!.x, accuracy: 1e-3, "\(j).x")
            XCTAssertEqual(mapped.y, live.points[j]!.y, accuracy: 1e-3, "\(j).y")
        }
    }

    func testPartialOverlapStillMaps() {
        // Live pose missing legs (upper body only) still yields a transform
        // that projects the FULL reference onto the body.
        let upper: [Joint] = [.nose, .leftEye, .rightEye, .neck, .leftShoulder,
                              .rightShoulder, .leftElbow, .rightElbow, .leftWrist,
                              .rightWrist, .root]
        let live = Fixtures.transformed(standing, scale: 2.0, translation: [1, 1])
        var partial: [Joint: SIMD2<Float>] = [:]
        for j in upper { partial[j] = live.points[j]! }
        let t = SimilarityTransform.mapping(reference: standing,
                                            live: PoseVector(points: partial))!
        // Check a LEG joint (absent from live) projects to where live legs are.
        let mappedAnkle = t.apply(standing.points[.leftAnkle]!)
        XCTAssertEqual(mappedAnkle.x, live.points[.leftAnkle]!.x, accuracy: 1e-3)
        XCTAssertEqual(mappedAnkle.y, live.points[.leftAnkle]!.y, accuracy: 1e-3)
    }

    func testTooFewJointsReturnsNil() {
        let tiny = PoseVector(points: [.nose: [0.5, 0.1], .neck: [0.5, 0.2]])
        XCTAssertNil(SimilarityTransform.mapping(reference: standing, live: tiny))
    }

    func testDegenerateReferenceReturnsNil() {
        var p: [Joint: SIMD2<Float>] = [:]
        for j in Joint.allCases { p[j] = [0.5, 0.5] }
        XCTAssertNil(SimilarityTransform.mapping(reference: PoseVector(points: p),
                                                 live: standing))
    }
}
